### ISSUE TITLE
Infrastructure Index (such as Unique Index on VersionInfo table) is not using IConventionSet

### ISSUE THREAD

**Issue by rodrigoramos** [fluentmigrator/fluentmigrator#1194]
I have customized my FluentMigrator project using IConventionSet but it doesn't seems to be considered when creating the infrastructure object (e.g. VersionInfo table's indexes).

**Example**

I have the following class correctly configured in my project:
``` C#
public class ProjectConstraintNameConvention : IConstraintConvention
{
    // Rest of code was omitted

    private static string GetConstraintName(ConstraintDefinition expression)
    {
        var sb = new StringBuilder();
        sb.Append("MYProject_");
        sb.Append(expression.IsPrimaryKeyConstraint ? "PK_" : "UC_");


        sb.Append(expression.TableName);
        foreach (var column in expression.Columns)
        {
            sb.Append('_').Append(column);
        }

        return sb.ToString();
    }
}
``` 

Running "Up" database, the following SQL is create:
``` SQL
CREATE UNIQUE INDEX UC_Version ON VersionInfo (Version ASC);
```

Instead of:

``` SQL
CREATE UNIQUE INDEX MYProject_UC_VersionInfo_Version ON VersionInfo (Version ASC);
```

I'm using version FluentMigrator/3.2.1

**Comment by tgharold** [comment]
Is this a .NET Core project where you are using `AddFluentMigratorCore()`?  

If so, you need to register the `IConventionSet` *after* making the  `AddFluentMigratorCore()` call in `Startup.ConfigureServices()`.  See [this sample for an example](https://github.com/tgharold/OptionsPatternValidation/blob/c4fcc6c5a1aaebcfac3c642c7d7c939bc5c7e408/examples/OptionsPatternMvc.Dapper/Extensions/ServiceCollectionExtensions.cs#L33-L44).

The reason for this (I think) is a line inside of `AddFluentMigratorCore()`.  Because the IServicesCollection is LIFO (last in first out), if `AddFluentMigratorCore()` registers something to handle `IConventionSet` after you register it... the FM registration will trump your registration.  The workaround is to register your `IConventionSet` after the FM registration.

Source:

https://github.com/fluentmigrator/fluentmigrator/blob/27eb0110ff14b4d32aed7d316ee5caf4dda20716/src/FluentMigrator.Runner/FluentMigratorServiceCollectionExtensions.cs#L101-L102

**Comment by jzabroski** [comment]
@tgharold Thanks for your help.  @rodrigoramos Let me know if we can close this issue.

**Comment by rodrigoramos** [comment]
Thanks for you answer.

I'm indeed using .NET Core. But the `IConventionSet` service is already registered after the `AddFluentMigratorCore()`.

That's my current configuration code:
``` C#
var serviceCollection = new ServiceCollection()
    .AddFluentMigratorCore()
    .ConfigureRunner(rb => rb
        .AddOracle12CManaged()
        .WithGlobalConnectionString(connStr)
        .ScanIn(typeof(ProjectVersionInfoTable).Assembly).For.VersionTableMetaData()
        .ScanIn(typeof(TabelaMarcadores).Assembly).For.Migrations()
    )
    .Configure<FluentMigratorLoggerOptions>(x => x.ShowSql = true)
    .AddLogging(lb => lb.AddFluentMigratorConsole())

``` 


A workaround that I have found was to override the `DefaultVersionTableMetaData` class as its `UniqueIndexName` property. That way, I was able to change the constraint name to whatever I liked it.

**Comment by jzabroski** [comment]
@rodrigoramos Hmm, I see.  Yes, this should still be regarded as a bug and fixed.  Thanks for the workaround!

**Comment by tgharold** [comment]
If the bug is that AddFluentMigratorCore() is registering a 2nd instance when an IConventionSet has already been registered, I do have a branch which will use `TryAddScoped()` instead of `AddScoped()` for that line.

**Comment by jzabroski** [comment]
Ah!  Thank you.  I was looking for the equivalent of Autofac's `IfNotRegistered()`.  Sweet sweet mother, what an obtuse name (`TryAddScoped()`) compared to Autofac's obvious one.  Granted, `IfNotRegistered()` is new in Autofac 4.4.  I don't see it documented in the FAQ under [How do I conditionally register components](https://autofaccn.readthedocs.io/en/latest/faq/conditional-registration.html), but the point stands: Autofac is way nicer syntax than Microsoft DI.

```c#
using Autofac.Core;

public class MyModule : Autofac.Module
{
  private readonly Assembly _migrationProjectAssembly;
  
  protected MyModule() : this(null) { }

  protected MyModule(Assembly migrationProjectAssembly)
  {
     _migrationProjectAssembly = migrationProjectAssembly;
  }

  protected override void Load(ContainerBuilder builder)
  {
    base.Load(builder);

    builder.RegisterType<Service>()
      .IfNotRegistered(typeof(IService))
      .AsImplementedInterfaces()
      .InstancePerLifetimeScope();
  }
}
```
@rodrigoramos Can you please try that and let me know how you make out?

**Comment by tgharold** [comment]
Yes, I pulled [TryAddScoped out of the MS documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1#service-registration-methods) while doing some other work.

> Generally, this method is used by library authors to avoid registering two copies of an instance in the container.

**Comment by jzabroski** [comment]
This was resolved awhile ago in 3.2.7.
